src/counter.tsx

```diff
@@ -0,0 +1,214 @@
+
+    React.useEffect(() => setCurrentValue(getValueInRange(defaultValue)), [defaultValue, getValueInRange]);
+
+    const updateValue = (newValue: number) => {
+        if (!isControlledByParent) {
+            setCurrentValue(newValue);
+        }
+        if (onChange) {
+            onChange(newValue);
+        }
+    };
+
+    return [isControlledByParent ? getValueInRange(value) : currentValue, updateValue];
+};
+
+type Props = {
+    value?: number;
+    defaultValue?: number;
+    min?: number;
+    max?: number;
+    onChange?: (value: number) => void;
+    onRemove?: () => void;
+    children?: void;
+    disabled?: boolean;
+    dataAttributes?: DataAttributes;
+    removeLabel?: string;
+    increaseLabel?: string;
+    decreaseLabel?: string;
+    valueLabel?: string;
+};
+
+const Counter: React.FC<Props> = ({
+    value,
+    defaultValue,
+    onChange,
+    onRemove,
+    min,
+    max,
+    dataAttributes,
+    disabled,
+    removeLabel,
+    increaseLabel,
+    decreaseLabel,
+    valueLabel,
+}) => {
+    const variant = useThemeVariant();
+    const {texts} = useTheme();
+
+    const minValue = min === undefined ? 0 : min;
+    const maxValue = Math.max(minValue, max === undefined ? 999 : max);
+
+    const [currentValue, setCurrentValue] = useCounterState({
+        value,
+        defaultValue,
+        min: minValue,
+        max: maxValue,
+        onChange,
+    });
+
+    const hasTrashIcon = !!onRemove && currentValue === minValue;
+
+    const getRemoveLabel = () => {
+        return removeLabel === undefined ? texts.counterRemoveLabel : removeLabel;
+    };
+
+    const getIncreaseLabel = () => {
+        return increaseLabel === undefined ? texts.counterIncreaseLabel : increaseLabel;
+    };
+
+    const getDecreaseLabel = () => {
+        return decreaseLabel === undefined ? texts.counterDecreaseLabel : decreaseLabel;
+    };
+
+    const getValueLabel = () => {
+        return `${currentValue}, ${valueLabel === undefined ? texts.counterQuantity : valueLabel}${
+            min !== undefined ? `, ${texts.counterMinValue} ${min}` : ''
+        }${max !== undefined ? `, ${texts.counterMaxValue} ${max}` : ''}`;
+    };
+
+    return (
+        <div
+            className={classNames(styles.counter, {[styles.disabled]: disabled})}
+            {...getPrefixedDataAttributes(dataAttributes, 'Counter')}
+            style={{
+                border: `1px solid ${
+                    variant === 'inverse' ? vars.colors.backgroundContainer : vars.colors.border
+                }`,
+            }}
+        >
+            <Inline space={8} alignItems="center">
+                <div
+                    className={classNames(styles.buttonContainer, {
+                        [styles.hasTrashIcon]: hasTrashIcon,
+                        [styles.isButtonDisabled]: disabled || (!hasTrashIcon && currentValue === minValue),
+                        [styles.disabled]: !disabled && !hasTrashIcon && currentValue === minValue,
```

Comment:
this logic is weird to me, are you applying a disabled class when `disabled` prop is false?

Reply:
yes, because there are two possible disabled states. One when the value is equal to min/max (only disable button), and another one when the entire counter is disabled (everything has less opacity, including the value and container). In the second case, if the value equals min/max, we don't want to apply the opacity filter twice to buttons